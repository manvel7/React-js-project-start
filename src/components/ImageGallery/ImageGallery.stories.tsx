import React from "react"
import {Meta, Story} from "@storybook/react/types-6-0"
import ImageGallery, {IProps} from "./ImageGallery"

const meta: Meta<IProps> = {
  title: "Components/ImageGallery",
  component: ImageGallery,
  argTypes: {}
}

const images = [
  "https://picsum.photos/id/11/1280/720",
  "https://picsum.photos/id/12/1024/768",
  "https://picsum.photos/id/13/1024/720",
  "https://picsum.photos/id/14/800/600",
  "https://picsum.photos/id/15/1024/768",
  "https://picsum.photos/id/16/1024/720",
  "https://picsum.photos/id/17/800/600",
  "https://picsum.photos/id/18/800/600",
  "https://picsum.photos/id/19/800/600",
  "https://picsum.photos/id/20/800/600"
]

const ImageGalleryStory: Story<IProps> = (args) => (
  <ImageGallery {...args} images={images} />
)

const Default = ImageGalleryStory.bind({})
Default.args = {}

export {Default}

export default meta
