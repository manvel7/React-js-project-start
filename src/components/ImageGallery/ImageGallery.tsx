import React, {
  FunctionComponent,
  HTMLAttributes,
  useMemo,
  useState
} from "react"
import ImageWrapper from "./ImageWrapper"
import ImageGalleryWrapper from "./Wrapper"
import CurrentImage from "./CurrentImage"
import ImageGalleryList from "./List"

const DEFAULT_HEIGHT = 40
const MIN_IMAGE_HEIGHT = 12

interface IImage {
  src: string
  index: number
}

export interface IProps extends HTMLAttributes<HTMLDivElement> {
  images: string[]
  containerHeight?: number
  containerWidth?: number
}

const ImageGallery: FunctionComponent<IProps> = (props) => {
  const {images, containerHeight, ...rest} = props

  const [active, setActive] = useState<number>(0)
  const [moreRight, setMoreRight] = useState<number>(0)
  const [moreLeft, setMoreLeft] = useState<number>(0)

  const imagesData = useMemo<IImage[]>(() => {
    return images.map((image, index) => ({index, src: image}))
  }, [images])

  const visibleImagesCount = useMemo<number>(() => {
    const count = Math.floor(
      (props.containerHeight || DEFAULT_HEIGHT) / MIN_IMAGE_HEIGHT
    )

    return count > 1 ? count : 2
  }, [props.containerHeight])

  const visibleImages = useMemo<IImage[]>(() => {
    let sliceFrom = 0

    if (active > imagesData.length - Math.ceil(visibleImagesCount / 2)) {
      sliceFrom = imagesData.length - visibleImagesCount
    } else if (active > 1) {
      sliceFrom = active - Math.floor(visibleImagesCount / 2)
    }

    if (sliceFrom < 0) {
      sliceFrom = 0
    }

    setMoreRight(imagesData.length - sliceFrom - visibleImagesCount)
    setMoreLeft(sliceFrom)

    return imagesData.slice(sliceFrom, sliceFrom + visibleImagesCount)
  }, [imagesData, visibleImagesCount, active])

  const activeImage = useMemo<IImage | undefined>(() => {
    return imagesData.find((image) => image.index === active)
  }, [imagesData, active])

  const selectImage = (image: IImage) => {
    setActive(image.index)
  }

  const onClickPrev = () => {
    if (active > 0) {
      setActive(active - 1)
    } else {
      setActive(imagesData.length - 1)
    }
  }

  const onClickNext = () => {
    if (active < imagesData.length - 1) {
      setActive(active + 1)
    } else {
      setActive(0)
    }
  }

  return (
    <ImageGalleryWrapper {...rest}>
      <CurrentImage
        src={activeImage?.src}
        containerHeight={containerHeight || DEFAULT_HEIGHT}
        onClickPrev={onClickPrev}
        onClickNext={onClickNext}
      />
      <ImageGalleryList moreRight={moreRight} moreLeft={moreLeft}>
        {visibleImages.map((image) => (
          <ImageWrapper
            key={image.index}
            isActive={active === image.index}
            onClick={() => selectImage(image)}>
            <img src={image.src} alt="" />
          </ImageWrapper>
        ))}
      </ImageGalleryList>
    </ImageGalleryWrapper>
  )
}

ImageGallery.defaultProps = {
  containerHeight: DEFAULT_HEIGHT
}

export default ImageGallery
