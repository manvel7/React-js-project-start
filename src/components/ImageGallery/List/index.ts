import ImageGalleryList from "./ImageGalleryList"

export default ImageGalleryList
