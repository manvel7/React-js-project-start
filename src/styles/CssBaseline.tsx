import {
  createGlobalStyle,
  css,
  DefaultTheme,
  GlobalStyleComponent,
  StyledProps
} from "styled-components"
import {mediaBreakpointDown} from "../utils/breakpoint"
import {getValue, toPercent} from "../utils/unit"
import {
  bodyMixin,
  smallMixin,
  h1Mixin,
  h2Mixin,
  h3Mixin,
  h4Mixin,
  h5Mixin,
  h6Mixin
} from "./typography"

interface IProps {}

const genericResetMixin = css<StyledProps<IProps>>`
  html,
  body,
  div,
  span,
  applet,
  object,
  iframe,
  h1,
  h2,
  h3,
  h4,
  h5,
  h6,
  p,
  blockquote,
  pre,
  a,
  abbr,
  acronym,
  address,
  big,
  cite,
  code,
  del,
  dfn,
  em,
  img,
  ins,
  kbd,
  q,
  s,
  samp,
  small,
  strike,
  strong,
  sub,
  sup,
  tt,
  var,
  b,
  u,
  i,
  center,
  dl,
  dt,
  dd,
  ol,
  ul,
  li,
  fieldset,
  form,
  label,
  legend,
  table,
  caption,
  tbody,
  tfoot,
  thead,
  tr,
  th,
  td,
  article,
  aside,
  canvas,
  details,
  embed,
  figure,
  figcaption,
  footer,
  header,
  hgroup,
  menu,
  nav,
  output,
  ruby,
  section,
  summary,
  time,
  mark,
  audio,
  video {
    margin: 0;
    padding: 0;
    border: 0;
    vertical-align: baseline;
  }

  article,
  aside,
  details,
  figcaption,
  figure,
  footer,
  header,
  hgroup,
  menu,
  nav,
  section {
    display: block;
  }

  ol,
  ul {
    list-style: none;
  }

  blockquote,
  q {
    quotes: none;
  }

  blockquote:before,
  blockquote:after,
  q:before,
  q:after {
    content: "";
    content: none;
  }

  table {
    border-collapse: collapse;
    border-spacing: 0;
  }
`

const appResetMixin = css<StyledProps<IProps>>`
  html {
    /* stylelint-disable property-no-vendor-prefix */
    -ms-overflow-style: scrollbar; /* Set box-sizing so that the width is not affected by padding or border */
    -ms-text-size-adjust: 100%; /* Set default font family in all browsers. */
    -webkit-tap-highlight-color: rgba(0, 0, 0, 0); /* Makes 1rem = 10px */
    -webkit-text-size-adjust: 100%; /* Prevent adjustments of font size after orientation changes in IE on Windows Phone and in iOS. */
    /* stylelint-enable property-no-vendor-prefix */
    box-sizing: border-box;
  }

  *,
  *::before,
  *::after {
    box-sizing: inherit;
  }

  body {
    background-color: ${({theme}) => theme.palette.background.white};

    /* Hide overflow content */
    &.overflow {
      overflow: hidden;
    }

    /* Prevent focus outline on elements that can't be accessed via keyboard. */
    [tabindex="-1"]:focus {
      outline: none !important;
    }

    /* Hide overflow in IE */
    svg:not(:root) {
      overflow: hidden;
    }

    /* Hide elements has 'hidden' attribute in IE 10- */
    [hidden] {
      display: none !important;
    }

    ::-webkit-file-upload-button {
      /* stylelint-disable-next-line property-no-vendor-prefix */
      -webkit-appearance: button;
      font: inherit;
    }
  }
`

const typographyMixin = css<StyledProps<IProps>>`
  html {
    font-family: ${({theme}) =>
      theme.typography
        .fontFamily}; /* Prevent scrollbars to overlap content in IE11 and Edge */
    font-size: ${({theme}) =>
      toPercent(
        getValue(theme.typography.htmlFontSize),
        16
      )}; /* Set default tap highlight to be transparent in iOS. */
    font-style: normal;
    color: ${({theme}) => theme.palette.text.dark};
  }

  body {
    ${bodyMixin<IProps>()}
  }

  h1 {
    ${h1Mixin<IProps>()}
  }

  h2 {
    ${h2Mixin<IProps>()}
  }

  h3 {
    ${h3Mixin<IProps>()}
  }

  h4 {
    ${h4Mixin<IProps>()}
  }

  h5 {
    ${h5Mixin<IProps>()}
  }

  h6 {
    ${h6Mixin<IProps>()}
  }

  small {
    ${smallMixin<IProps>()}
  }

  b,
  strong {
    font-weight: ${({theme}) => theme.typography.fontWeights.bold};

    h1,
    h2,
    h3,
    h4,
    h5,
    h6,
    caption {
      font-weight: ${({theme}) => theme.typography.fontWeights.boldest};
    }
  }
`

const utilsMixin = css<StyledProps<IProps>>`
  .hidden {
    visibility: hidden;
  }

  .d-none {
    display: none;
  }

  .hidden-sm-down {
    ${({theme}) =>
      mediaBreakpointDown(
        "sm",
        theme.grid.breakpoints,
        `
      display: none;
    `
      )}
  }

  .text-center {
    text-align: center;
  }
`

const CssBaseline: GlobalStyleComponent<
  IProps,
  DefaultTheme
> = createGlobalStyle`
  ${genericResetMixin}

  ${appResetMixin}

  ${typographyMixin}

  ${utilsMixin}
`

export default CssBaseline
