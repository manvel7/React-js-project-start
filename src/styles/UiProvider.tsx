import React, {FunctionComponent} from "react"
import {ThemeProvider} from "styled-components"
import theme from "./theme"
import {CssBaseline} from "./index"

interface IProps {}

const UiProvider: FunctionComponent<IProps> = (props) => {
  const {children} = props

  return (
    <ThemeProvider theme={theme}>
      <CssBaseline />
      {children}
    </ThemeProvider>
  )
}

export default UiProvider
