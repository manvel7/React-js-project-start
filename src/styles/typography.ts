import {
  css,
  DefaultTheme,
  FlattenInterpolation,
  StyledProps,
  ThemedStyledProps
} from "styled-components"
import {pxToRem} from "../utils/unit"

const h1Mixin = <P extends Record<any, any>>(): FlattenInterpolation<
  ThemedStyledProps<P, DefaultTheme>
> => css<StyledProps<P>>`
  font-size: ${({theme}) =>
    pxToRem(theme.typography.variants.h1.size, theme.typography.htmlFontSize)};
  font-weight: ${({theme}) => theme.typography.variants.h1.weight};
  letter-spacing: ${({theme}) => theme.typography.variants.h1.letterSpacing};
  line-height: ${({theme}) =>
    pxToRem(
      theme.typography.variants.h1.lineHeight,
      theme.typography.htmlFontSize
    )};

  b,
  strong {
    font-weight: ${({theme}) => theme.typography.fontWeights.boldest};
  }
`

const h2Mixin = <P extends Record<any, any>>(): FlattenInterpolation<
  ThemedStyledProps<P, DefaultTheme>
> => css<StyledProps<P>>`
  font-size: ${({theme}) =>
    pxToRem(theme.typography.variants.h2.size, theme.typography.htmlFontSize)};
  font-weight: ${({theme}) => theme.typography.variants.h2.weight};
  letter-spacing: ${({theme}) => theme.typography.variants.h2.letterSpacing};
  line-height: ${({theme}) =>
    pxToRem(
      theme.typography.variants.h2.lineHeight,
      theme.typography.htmlFontSize
    )};

  b,
  strong {
    font-weight: ${({theme}) => theme.typography.fontWeights.boldest};
  }
`

const h3Mixin = <P extends Record<any, any>>(): FlattenInterpolation<
  ThemedStyledProps<P, DefaultTheme>
> => css<StyledProps<P>>`
  font-size: ${({theme}) =>
    pxToRem(theme.typography.variants.h3.size, theme.typography.htmlFontSize)};
  font-weight: ${({theme}) => theme.typography.variants.h3.weight};
  letter-spacing: ${({theme}) => theme.typography.variants.h3.letterSpacing};
  line-height: ${({theme}) =>
    pxToRem(
      theme.typography.variants.h3.lineHeight,
      theme.typography.htmlFontSize
    )};

  b,
  strong {
    font-weight: ${({theme}) => theme.typography.fontWeights.boldest};
  }
`

const h4Mixin = <P extends Record<any, any>>(): FlattenInterpolation<
  ThemedStyledProps<P, DefaultTheme>
> => css<StyledProps<P>>`
  font-size: ${({theme}) =>
    pxToRem(theme.typography.variants.h4.size, theme.typography.htmlFontSize)};
  font-weight: ${({theme}) => theme.typography.variants.h4.weight};
  letter-spacing: ${({theme}) => theme.typography.variants.h4.letterSpacing};
  line-height: ${({theme}) =>
    pxToRem(
      theme.typography.variants.h4.lineHeight,
      theme.typography.htmlFontSize
    )};

  b,
  strong {
    font-weight: ${({theme}) => theme.typography.fontWeights.boldest};
  }
`

const h5Mixin = <P extends Record<any, any>>(): FlattenInterpolation<
  ThemedStyledProps<P, DefaultTheme>
> => css<StyledProps<P>>`
  font-size: ${({theme}) =>
    pxToRem(theme.typography.variants.h5.size, theme.typography.htmlFontSize)};
  font-weight: ${({theme}) => theme.typography.variants.h5.weight};
  letter-spacing: ${({theme}) => theme.typography.variants.h5.letterSpacing};
  line-height: ${({theme}) =>
    pxToRem(
      theme.typography.variants.h5.lineHeight,
      theme.typography.htmlFontSize
    )};

  b,
  strong {
    font-weight: ${({theme}) => theme.typography.fontWeights.boldest};
  }
`

const h6Mixin = <P extends Record<any, any>>(): FlattenInterpolation<
  ThemedStyledProps<P, DefaultTheme>
> => css<StyledProps<P>>`
  font-size: ${({theme}) =>
    pxToRem(theme.typography.variants.h6.size, theme.typography.htmlFontSize)};
  font-weight: ${({theme}) => theme.typography.variants.h6.weight};
  letter-spacing: ${({theme}) => theme.typography.variants.h6.letterSpacing};
  line-height: ${({theme}) =>
    pxToRem(
      theme.typography.variants.h6.lineHeight,
      theme.typography.htmlFontSize
    )};

  b,
  strong {
    font-weight: ${({theme}) => theme.typography.fontWeights.boldest};
  }
`

const bodyMixin = <P extends Record<any, any>>(): FlattenInterpolation<
  ThemedStyledProps<P, DefaultTheme>
> => css<StyledProps<P>>`
  font-size: ${({theme}) =>
    pxToRem(
      theme.typography.variants.body.size,
      theme.typography.htmlFontSize
    )};
  font-weight: ${({theme}) => theme.typography.variants.body.weight};
  letter-spacing: ${({theme}) => theme.typography.variants.body.letterSpacing};
  line-height: ${({theme}) =>
    pxToRem(
      theme.typography.variants.body.lineHeight,
      theme.typography.htmlFontSize
    )};
`

const smallMixin = <P extends Record<any, any>>(): FlattenInterpolation<
  ThemedStyledProps<P, DefaultTheme>
> => css<StyledProps<P>>`
  font-size: ${({theme}) =>
    pxToRem(
      theme.typography.variants.caption.size,
      theme.typography.htmlFontSize
    )};
  font-weight: ${({theme}) => theme.typography.variants.caption.weight};
  letter-spacing: ${({theme}) =>
    theme.typography.variants.caption.letterSpacing};
  line-height: ${({theme}) =>
    pxToRem(
      theme.typography.variants.caption.lineHeight,
      theme.typography.htmlFontSize
    )};

  b,
  strong {
    font-weight: ${({theme}) => theme.typography.fontWeights.boldest};
  }
`
export {
  h1Mixin,
  h2Mixin,
  h3Mixin,
  h4Mixin,
  h5Mixin,
  h6Mixin,
  bodyMixin,
  smallMixin
}
