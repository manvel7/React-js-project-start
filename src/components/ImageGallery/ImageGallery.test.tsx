import React from "react"
import {render} from "@testing-library/react"
import ImageGallery from "./ImageGallery"
import {UiProvider} from "../../styles"

const images = [
  "https://picsum.photos/id/11/1280/720",
  "https://picsum.photos/id/12/1024/768",
  "https://picsum.photos/id/13/1024/720"
]

describe("ImageGallery", () => {
  describe("Default", () => {
    it("should match snapshot", () => {
      const result = render(
        <UiProvider>
          <ImageGallery images={images} />
        </UiProvider>
      )
      expect(result.container.firstChild).toMatchSnapshot()
    })
  })
})
