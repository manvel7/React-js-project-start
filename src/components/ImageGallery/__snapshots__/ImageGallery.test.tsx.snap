// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageGallery Default should match snapshot 1`] = `
.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 0 1 25%;
  -ms-flex: 0 1 25%;
  flex: 0 1 25%;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c4 > div {
  margin-bottom: 0.8rem;
}

.c4 > div:last-child {
  margin-bottom: 0;
}

.c2 {
  border: 0.1rem solid transparent;
  border-radius: 0.6rem;
  overflow: hidden;
}

.c2 img {
  height: 100%;
  object-fit: cover;
  vertical-align: middle;
  width: 100%;
}

.c3 .c2 {
  cursor: pointer;
  -webkit-flex: 1 1 0;
  -ms-flex: 1 1 0;
  flex: 1 1 0;
}

.c5 {
  border: 0.1rem solid transparent;
  border-radius: 0.6rem;
  overflow: hidden;
  border-color: #001254;
}

.c5 img {
  height: 100%;
  object-fit: cover;
  vertical-align: middle;
  width: 100%;
}

.c3 .c5 {
  cursor: pointer;
  -webkit-flex: 1 1 0;
  -ms-flex: 1 1 0;
  flex: 1 1 0;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  width: 100%;
}

.c1 {
  -webkit-flex: 1 0 75%;
  -ms-flex: 1 0 75%;
  flex: 1 0 75%;
  height: 40rem;
  margin-right: 0.8rem;
  position: relative;
}

.c1 .nav-buttons {
  background-color: transparent;
  bottom: 1.6rem;
  left: 1.6rem;
  position: absolute;
}

.c1 .nav-buttons button {
  background-color: #FEFEFE;
  border: 0.1rem solid #E6E6E6;
  border-radius: 0.4rem;
  color: #001254;
  cursor: pointer;
  height: 4rem;
  width: 4rem;
}

.c1 .nav-buttons button:first-child {
  margin-right: 0.8rem;
}

@media (max-width:575px) {
  .c4 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
    margin-top: 0.8rem;
    height: 8rem;
  }

  .c4 > div {
    margin-bottom: 0rem;
    margin-right: 0.8rem;
  }

  .c4 > div:last-child {
    margin-right: 0;
  }
}

@media (max-width:575px) {
  .c0 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

@media (max-width:575px) {
  .c1 {
    margin-right: 0;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
      style="height: 100%;"
    >
      <img
        alt=""
        src="https://picsum.photos/id/11/1280/720"
      />
    </div>
    <div
      class="nav-buttons hidden-sm-down"
    >
      <button>
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-arrow-left fa-w-14 fa-lg "
          data-icon="arrow-left"
          data-prefix="far"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M229.9 473.899l19.799-19.799c4.686-4.686 4.686-12.284 0-16.971L94.569 282H436c6.627 0 12-5.373 12-12v-28c0-6.627-5.373-12-12-12H94.569l155.13-155.13c4.686-4.686 4.686-12.284 0-16.971L229.9 38.101c-4.686-4.686-12.284-4.686-16.971 0L3.515 247.515c-4.686 4.686-4.686 12.284 0 16.971L212.929 473.9c4.686 4.686 12.284 4.686 16.971-.001z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button>
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-arrow-right fa-w-14 fa-lg "
          data-icon="arrow-right"
          data-prefix="far"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M218.101 38.101L198.302 57.9c-4.686 4.686-4.686 12.284 0 16.971L353.432 230H12c-6.627 0-12 5.373-12 12v28c0 6.627 5.373 12 12 12h341.432l-155.13 155.13c-4.686 4.686-4.686 12.284 0 16.971l19.799 19.799c4.686 4.686 12.284 4.686 16.971 0l209.414-209.414c4.686-4.686 4.686-12.284 0-16.971L235.071 38.101c-4.686-4.687-12.284-4.687-16.97 0z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
  </div>
  <div
    class="c3 c4"
  >
    <div
      class="c5"
    >
      <img
        alt=""
        src="https://picsum.photos/id/11/1280/720"
      />
    </div>
    <div
      class="c2"
    >
      <img
        alt=""
        src="https://picsum.photos/id/12/1024/768"
      />
    </div>
    <div
      class="c2"
    >
      <img
        alt=""
        src="https://picsum.photos/id/13/1024/720"
      />
    </div>
  </div>
</div>
`;
